Pull 202405 Commits into 202502 [REBASE&FF]

apop5 opened the pull request:
## Description

Pull in the commits from 202405 Branch to 202502 Branch.

- [ ] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [ ] Includes documentation?
- [ ] Backport to release branch?

## How This Was Tested
CI Passes with changes pulled.

## Integration Instructions
No integration required.

makubacki committed:
[CHERRY-PICK] MdePkg/BaseMemoryLib: Prevent potential VS2022 linker failure

After updating between various VS2022 versions such as 17.4 to
17.5, , linker failures began to appear in several modules because
`memset` is an unresolved symbol.

The following functions in BaseMemoryLib/MemLibGeneric.c have their
loop pattern replaced with the `memset` intrinsic function:

- `InternalMemSetMem16()`
- `InternalMemSetMem32()`
- `InternalMemSetMem64()`

An example of an error related to `InternalMemSetMem64()` in
VariableSmmRuntimeDxe is shown below:

```
INFO - BaseMemoryLib.lib(MemLibGeneric.obj) : error LNK2001:
         unresolved external symbol memset
INFO - <...>\VariableSmmRuntimeDxe.dll : fatal error LNK1120:
         1 unresolved externals
```

This was reproduced in several environments including:

- Public VM image:
  https://github.com/actions/runner-images/blob/win22/20230226.1/images/win/Windows2022-Readme.md

- Locally when updating from 17.4.4 to 17.5.1

> Note: This image (with 17.4) does not have this issue
  https://github.com/actions/runner-images/blob/win22/20230219.1/images/win/Windows2022-Readme.md

This change updates the type cast for the destination buffer to be
a pointer to `volatile` data to prevent this optimization with a
relatively minimum delta to prior implementation.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>
(cherry picked from commit c46bc0ea98b4a8483c970e2282ea46891ea94f57)

kuqin12 committed:
BaseTools: Add GenStm

Adds GenStm as a C tool.

Co-authored-by: Michael Kubacki <michael.kubacki@microsoft.com>
Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

makubacki committed:
[CHERRY-PICK] MdeModulePkg: Add Standalone MM Core performance support

Adds a new library instance to support MM core functionality for
performance in Standalone MM.

- Add StandaloneMmCorePerformanceLib instance
- Move common MM logic to a new file `MmCorePerformanceLib.c`
- Define interfaces with implementation specific to MM environment
  type in `SmmCorePerformanceLibInternal.h` and implement those
  functions in the Standalone MM and Traditional MM specific C files

Note: StandaloneMmCorePerformanceLib supports both
      `MM_CORE_STANDALONE` and `MM_STANDALONE` as some Standalone MM
	  environments have privilege separation and need to link this
	  functionality in a ring 3 Standalone MM driver that is outside
	  the ring 0 Standalone MM core driver.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>
(cherry picked from commit 11b44c5cd15e3bc83a68d42644faa30d3016a7b4)

makubacki committed:
[CHERRY-PICK] MdeModulePkg: Add Standalone MM perf library support

Adds a new library instance to support logging performance data in
Standalone MM.

- Add StandaloneMmPerformanceLib instance
- Move common MM logic to a new file `SmmPerformanceLibInternal.c`
- Since the library largely defers most logic to the performance
  measurement protocol a large degree of code can be shared between
  Standalone MM and Traditional MM.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>
(cherry picked from commit 1c51a268b717640e48c7c804d92b45301383e219)

makubacki committed:
[CHERRY-PICK] MdeModulePkg/HiiDatabaseDxe: Prevent linker error

Prevent an issue where `memcpy()` instrinsic is being used after
recent MSVC linker update in windows-2022 VM image from 14.31.31103
to 14.32.31326.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>
(cherry picked from commit 29f02d0161bc5d10a6e3b2cfc4d10799644964a5)

makubacki committed:
[CHERRY-PICK] DebugMacroCheck: Do not show progress bar with zero items

In the case that the total provided to the `_show_progress()`
function is zero, do not show a progress bar to prevent aborts
`ZeroDivisionError` when calculating the progress percentage.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>
(cherry picked from commit 67df979ad2ea0a5d344bd997a7b3b1e87c7e5cc8)

makubacki committed:
[CHERRY-PICK] MdeModulePkg/DxeCorePerformanceLib: Use EFI_MM_COMMUNICATE_HEADER

Refactor the file to use the new type name EFI_MM_COMMUNICATE_HEADER.

This is the same type but follows the new name and is more clear
with upcoming Standalone MM support being added.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>
(cherry picked from commit 874c24baae5e89ccb96d753b2ac69a7f5f9b2b68)

dependabot[bot] committed:
GitHub Action: Bump robinraju/release-downloader from 1.11 to 1.12 (#1295)

Bumps
[robinraju/release-downloader](https://github.com/robinraju/release-downloader)
from 1.11 to 1.12.

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Aaron <aaronpop@microsoft.com>

SeolforHsieh committed:
MdePkg: Add Google Mock Protocol for MmCommunication2 (#1288)

SeolforHsieh committed:
MdePkg: Add all mock functions in the MockMmServicesTableLib (#1279)

SeolforHsieh committed:
StandaloneMmPkg: Add Google Mock Library for StandaloneMmMemLib (#1287)

VivianNK committed:
MdePkg: Complete MockUefiBootServicesTableLib

This commit completes the MockUefiBootServicesTableLib by adding the
missing functions and data structures.

This includes MockSimpleTextIn and MockSimpleTextOut protocols for the
EFI_SYSTEM_TABLE.

dependabot[bot] committed:
pip: bump pygount from 1.8.0 to 2.0.0 (#1309)

Bumps [pygount](https://github.com/roskakori/pygount) from 1.8.0 to
2.0.0.

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

xuweiintel committed:
[CHERRY-PICK] MdeModulePkg/DxeCorePerformanceLib: Fix incorrect size calculation

The values of BootRecordDataPayloadSize and CommSize are incorrect.
BootRecordDataPayloadSize should equal to
  SmmBootRecordDataSize - SmmBootRecordDataRetrieved
CommSize should equal to
  OFFSET_OF (EFI_MM_COMMUNICATE_HEADER,Data) \
  + (UINTN)MmCommBufferHeader->MessageLength
SmmCommData->BootRecordSize should be set to BootRecordDataPayloadSize,
instead of the total size of entire Smm boot record data.

Signed-off-by: Wei6 Xu <wei6.xu@intel.com>
(cherry picked from commiit 95bf74fac1977a083234518933393939145160fc)

uefibot committed:
Repo File Sync: synced file(s) with microsoft/mu_devops (#1300)

synced local file(s) with
[microsoft/mu_devops](https://github.com/microsoft/mu_devops).

🤖: View the [Repo File Sync Configuration
File](https://github.com/microsoft/mu_devops/blob/main/.sync/Files.yml)
to see how files are synced.



---

This PR was created automatically by the
[repo-file-sync-action](https://github.com/BetaHuhn/repo-file-sync-action)
workflow run
[#14031159574](https://github.com/microsoft/mu_devops/actions/runs/14031159574)

Signed-off-by: Project Mu UEFI Bot <uefibot@microsoft.com>

makubacki committed:
[CHERRY-PICK] StandaloneMmCorePerformanceLib.inf: Add gEfiEventExitBootServicesGuid

Adds `gEfiEventExitBootServicesGuid` to the `[Guids]` section and
removes `gEdkiiSmmExitBootServicesProtocolGuid` from the
`[Protocols]` section for the current implementation.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>
(cherry picked from commit 30f3fa0af7bbe974bb4b2a401ff0db26c29a2e02)

uefibot committed:
RustToolChain: Bump RustToolChain from 1.80 to 1.84.

Signed-off-by: Project Mu UEFI Bot <uefibot@microsoft.com>

apop5 committed:
pip: bump edk2-pytool-library to 0.23.2, edk2-pytool-extensions to 0.29.2

kenlautner committed:
Add CPU_INFORMATION_HEADER struct to STM_HEADER

dependabot[bot] committed:
GitHub Action: Bump robinraju/release-downloader from 1.11 to 1.12 (#1318)

Bumps
[robinraju/release-downloader](https://github.com/robinraju/release-downloader)
from 1.11 to 1.12.

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

kenlautner committed:
Move STATIC_ASSERT abstraction to common file for BaseTools usage

apop5 committed:
Squash on Rebase: Clean up Edk2 CI files.

Squash with Enable CI on next rebase.

kenlautner commented on .azurepipelines/MuDevOpsWrapper.yml:
```diff
@@ -19,7 +19,7 @@ resources:
       type: github
```
I'm leaving this comment here so you need to resolve it before completing the PR.

I don't think all these commits should be squashed together.  I think the commit history needs to be cleaned up (and some things squashed together) but we should merge this PR as a Rebase and Fast-Forward.

makubacki committed:
BaseTools/UefiVarPatcher: Add plugin (#1326)

## Description

Adds a build plugin that runs in post-build to modify the UEFI variable
store in the ROM image based on variable data in an XML if the plugin is
enabled.

Details are in UefiVarPatcher/ReadMe.md.

- [ ] Impacts functionality?
- [ ] Impacts security?
- [ ] Breaking change?
- [ ] Includes tests?
- [x] Includes documentation?
- [x] Backport to release branch?

## How This Was Tested

- edk2-pytool-extensions unit tests
- Run plugin locally in mu_tiano_platforms
- Tested on a physical platform build

## Integration Instructions

- See `BaseTools/Plugin/UefiVarPatcher/ReadMe.md` for details

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

makubacki commented:
@apop5, it just went in, but could you please include this change?
https://github.com/microsoft/mu_basecore/pull/1326
